### PR TITLE
Module renderer

### PR DESCRIPTION
**Pull request by jwildfire**
# Overview

This branch refactors the process for adding custom charts and reimplements shiny module renderers. See the draft of the new [Custom Chart Vignette](https://github.com/SafetyGraphics/safetyGraphics/wiki/Chart-Configuration-Vignette-(v2)) for details about the new process. 

# Details

The biggest change is that chart configuration information has been migrated from `raw-data/charts.csv`  to a series of yaml files. The yaml files are currently saved in `inst/config/charts`, but will be migrated to the safetyCharts package in the near future. Detailed documentation of the YAML format is saved in this wiki page, which will be adapted in to a vignette before the v2 release. 

# Test notes

Checkout the branch and then run this code to start the app. 

```
devtools::install_github("safetyGraphics/safetyCharts")
library(safetyCharts)
devtools::install()
library(safetyGraphics)
safetyGraphics::safetyGraphicsApp()
```
Just load the package and run `safetyGraphicsApp()` like normal. You can also run `makeChartConfig()` directly to easily create a list of charts from the YAML files saved in a set of directories.

### CODE REVIEW

File: R/app_ui.R [hunk 1]
```diff
@@ -8,39 +8,45 @@
...
-        navbarMenu('Charts', icon=icon("chart-bar")),
-        tabPanel("Reports", icon=icon("file-alt")),
-        navbarMenu('',icon=icon("cog"),
-          tabPanel(title = "Metadata", settingsMappingUI("metaSettings")),
-          tabPanel(title = "Charts", settingsChartsUI("chartSettings"))
-        ),
-        tags$script(
-          HTML(
+    #read css from pacakge
+    app_css <-  HTML(readLines( paste(.libPaths(),'safetygraphics','safetyGraphics_app', 'www','index.css', sep="/")))
```
Comment (xni7) [inline R/app_ui.R:12]: @jwildfire, there is a typo in safetyGraphics, should be capital G.
```suggestion
    app_css <-  HTML(readLines( paste(.libPaths(),'safetyGraphics','safetyGraphics_app', 'www','index.css', sep="/")))
```

**Comment by xni7** [comment]
> # Test notes
> Checkout the branch and then run this code to start the app.
> 
> ```
> devtools::install_github("safetyGraphics/safetyCharts") 
> library(safetyCharts)
> devtools::install()
> library(safetyGraphics)
> safetyGraphics::safetyGraphicsApp()
> ```

I had issue installing devtools::install_github("safetyGraphics/safetyCharts"), it complained **"object 'nep_explorer_data' not found"**. There is some test code in R_old that requires this test data but not included

File: R/app_ui.R (hunk 1 above)
Comment (jwildfire) [reply R/app_ui.R]: Good catch. Not entirely sure the best way to deal with the libPaths issue, so filed #441 to look at it more closely later.

**Comment by jwildfire** [comment]
> There is some test code in R_old that requires this test data but not included.R_old that requires this test data but not included

I just added R_Old to .rbuildignore in safetycharts. I'm able to build ok here, but wasn't getting that bug before either, so let me know if that fixes the issue on your end.

**Comment by xni7** [comment]
> > There is some test code in R_old that requires this test data but not included.R_old that requires this test data but not included
> 
> @xni7 I just added R_Old to .rbuildignore in safetycharts. I'm able to build ok here, but wasn't getting that bug before either, so let me know if that fixes the issue on your end.

it installs okay now

**Comment by xni7** [comment]
@jwildfire , couldn't get [Example 1 - Static Outlier Explorer](https://github.com/SafetyGraphics/safetyGraphics/wiki/Chart-Configuration-Vignette-(v2)#example-1---static-outlier-explorer) working. Seems to have do with the ggplot aes mapping.
Also where to load user supplied custom data? Is this implemented yet?

image

**Comment by jwildfire** [comment]
> @jwildfire , couldn't get [Example 1 - Static Outlier Explorer](https://github.com/SafetyGraphics/safetyGraphics/wiki/Chart-Configuration-Vignette-(v2)#example-1---static-outlier-explorer) working. Seems to have do with the ggplot aes mapping.

Hm, seems like a data issue ... Is there any detail in the R log? I'm going to fix the data preview section now so that this will be easier to debug.

**Comment by jwildfire** [comment]
> Also where to load user supplied custom data? Is this implemented yet?

Right now, the workflow is to load your data in R and then pass it in to the safetyGraphics app like so: 

```
mylabsdata <- read.csv(...)

safetyGraphicsApp(data=list(labs=myLabdata, ...)
```

Eventually we want to create a helper app/plugin/widget that allows non-technical users to load data (and apply other customizations) via a simple GUI. This will likely look at least a little bit like Bo's shinymeta workflow. This is tracked in #427 so we can discuss details there.

**Comment by jwildfire** [comment]
> I'm going to fix the data preview section now so that this will be easier to debug.
 
Updated now. Can you check your labs data and confirm that it is showing up as expected and has a column called "PARAM"

**Comment by MayaGans** [comment]
> You can also run `makeChartsSettings()` directly to easily create a list of charts from the YAML files saved in a set of directories.

I was able to open the app (woohoo!) and I wanted to try this alternative you posed, @jwildfire but I don't see this function anywhere when I look using `safetyGraphics:::*` ?

**Comment by jwildfire** [comment]
> I don't see this function anywhere when I look using safetyGraphics:::* ?

@MayaGans It was a typo. I meant `makeChartConfig`. I'll edit the original message to fix in a sec. The workflow I was alluding was to call `makeChartConfig()` directly and then pass the results to `safetyGraphicsApp()` like this: 

```
customDirectories<- c("some/custom/charts", "safetyCharts/inst/config")
customCharts<-makeChartConfig(dirs=customDirectories)
safetyGraphicsApp(charts=customCharts)
```

@xni7 this is the current workflow for adding custom charts to the defaults found in safetyCharts.

**Comment by jwildfire** [comment]
Merging after discussion with technical team - Still plenty of work to do, but everyone is generally able to run the app locally! Onward!

Close #429 #431 #438 #442 #450

**Comment by jwildfire** [comment]
Oops - merging before closing ...

**Pull request merged**